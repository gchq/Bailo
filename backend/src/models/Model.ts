import { model, Schema } from 'mongoose'
import MongooseDelete, { SoftDeleteDocument } from 'mongoose-delete'

export const EntryVisibility = {
  Private: 'private',
  Public: 'public',
} as const

export type EntryVisibilityKeys = (typeof EntryVisibility)[keyof typeof EntryVisibility]

export const EntryKind = {
  Model: 'model',
  DataCard: 'data-card',
} as const

export type EntryKindKeys = (typeof EntryKind)[keyof typeof EntryKind]

export interface CollaboratorEntry {
  entity: string
  roles: Array<'owner' | 'contributor' | 'consumer' | string>
}

export interface ModelMetadata {
  overview?: {
    tags: Array<string>
    [x: string]: unknown
  }

  // allow other properties
  [x: string]: unknown
}

export interface ModelCardInterface {
  schemaId: string
  version: number
  createdBy: string

  metadata: ModelMetadata
}

export interface Settings {
  ungovernedAccess: boolean
  allowTemplating: boolean
  mirror: { sourceModelId?: string; destinationModelId?: string }
}

// This interface stores information about the properties on the base object.
// It should be used for plain object representations, e.g. for sending to the
// client.
export interface ModelInterface {
  id: string

  name: string
  kind: EntryKindKeys
  description: string
  card?: ModelCardInterface
  organisation: string
  state: string

  collaborators: Array<CollaboratorEntry>
  settings: Settings

  visibility: EntryVisibilityKeys
  deleted: boolean

  createdAt: Date
  updatedAt: Date
}

// The doc type includes all values in the plain interface, as well as all the
// properties and functions that Mongoose provides.  If a function takes in an
// object from Mongoose it should use this interface
export type ModelDoc = ModelInterface & SoftDeleteDocument

const ModelSchema = new Schema<ModelDoc>(
  {
    id: { type: String, required: true, unique: true, index: true },

    name: { type: String, required: true },
    kind: { type: String, enum: Object.values(EntryKind) },
    description: { type: String, required: true },
    card: {
      schemaId: { type: String },
      version: { type: Number },
      createdBy: { type: String },

      metadata: { type: Schema.Types.Mixed },
    },
    organisation: {
      type: String,
      required: function () {
        return typeof this['organisation'] === 'string' ? false : true
      },
      default: '',
    },
    state: {
      type: String,
      required: function () {
        return typeof this['state'] === 'string' ? false : true
      },
      default: '',
    },

    collaborators: [
      {
        entity: { type: String, required: true },
        roles: [{ type: String }],
      },
    ],
    settings: {
      ungovernedAccess: { type: Boolean, required: true, default: false },
      allowTemplating: { type: Boolean, required: true, default: false },
      mirror: { sourceModelId: { type: String }, destinationModelId: { type: String } },
    },

    visibility: { type: String, enum: Object.values(EntryVisibility), default: EntryVisibility.Public },
  },
  {
    timestamps: true,
    collection: 'v2_models',
  },
)

ModelSchema.plugin(MongooseDelete, {
  overrideMethods: 'all',
  deletedBy: true,
  deletedByType: Schema.Types.ObjectId,
  deletedAt: true,
})
ModelSchema.index({ '$**': 'text' }, { weights: { name: 10, description: 5 } })

const ModelModel = model<ModelDoc>('v2_Model', ModelSchema)

export default ModelModel
