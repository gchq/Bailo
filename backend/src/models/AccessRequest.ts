import { model, Schema } from 'mongoose'
import MongooseDelete, { SoftDeleteDocument } from 'mongoose-delete'

export interface AccessRequestMetadata {
  overview: {
    name: string
    entities: Array<string>

    endDate?: string
    [x: string]: unknown
  }

  [x: string]: unknown
}

// This interface stores information about the properties on the base object.
// It should be used for plain object representations, e.g. for sending to the
// client.
export interface AccessRequestInterface {
  id: string
  modelId: string

  schemaId: string
  metadata: AccessRequestMetadata

  deleted: boolean

  createdBy: string
  createdAt: Date
  updatedAt: Date
}

// The doc type includes all values in the plain interface, as well as all the
// properties and functions that Mongoose provides.  If a function takes in an
// object from Mongoose it should use this interface

export type AccessRequestDoc = AccessRequestInterface & SoftDeleteDocument

const AccessRequestSchema = new Schema<AccessRequestDoc>(
  {
    id: { type: String, unique: true, required: true },
    modelId: { type: String, required: true },

    schemaId: { type: String, required: true },
    metadata: { type: Schema.Types.Mixed, required: true },

    createdBy: { type: String, required: true },
  },
  {
    timestamps: true,
    collection: 'v2_access_requests',
  },
)

AccessRequestSchema.plugin(MongooseDelete, {
  overrideMethods: 'all',
  deletedBy: true,
  deletedByType: String,
  deletedAt: true,
})

const AccessRequestModel = model<AccessRequestDoc>('v2_Access_Request', AccessRequestSchema)

export default AccessRequestModel
