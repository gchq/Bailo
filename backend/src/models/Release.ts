import { Document, model, Schema } from 'mongoose'
import MongooseDelete from 'mongoose-delete'

import { semverObjectToString, semverStringToObject } from '../services/release.js'

// This interface stores information about the properties on the base object.
// It should be used for plain object representations, e.g. for sending to the
// client.
export interface ReleaseInterface {
  modelId: string
  modelCardVersion: number

  semver: string
  notes: string

  minor: boolean
  draft: boolean

  fileIds: Array<string>
  images: Array<ImageRef>

  deleted: boolean

  createdBy: string
  createdAt: Date
  updatedAt: Date
}

export interface ImageRef {
  repository: string
  name: string
  tag: string
}

// The doc type includes all values in the plain interface, as well as all the
// properties and functions that Mongoose provides.  If a function takes in an
// object from Mongoose it should use this interface
export type ReleaseDoc = ReleaseInterface & Document<any, any, ReleaseInterface>

export interface SemverObject {
  major: number
  minor: number
  patch: number
  metadata?: string
}

const ReleaseSchema = new Schema<ReleaseInterface & { semver: string | SemverObject }>(
  {
    modelId: { type: String, required: true },
    modelCardVersion: { type: Number, required: true },

    semver: {
      type: Schema.Types.Mixed,
      required: true,
      set: function (semver: string) {
        return semverStringToObject(semver)
      },
      get: function (semver: SemverObject | string) {
        if (typeof semver === 'string') {
          return semver
        } else return semverObjectToString(semver)
      },
    },

    notes: { type: String, required: true },

    minor: { type: Boolean, required: true },
    draft: { type: Boolean, required: true },

    fileIds: [{ type: Schema.Types.ObjectId }],
    images: [
      {
        repository: { type: String },
        name: { type: String },
        tag: { type: String },
      },
    ],

    createdBy: { type: String, required: true },
  },
  {
    timestamps: true,
    collection: 'v2_releases',
    toJSON: { getters: true },
    toObject: { getters: true },
  },
)

ReleaseSchema.plugin(MongooseDelete, { overrideMethods: 'all', deletedBy: true, deletedByType: Schema.Types.ObjectId })
ReleaseSchema.index({ modelId: 1, semver: 1 }, { unique: true })

const ReleaseModel = model<ReleaseInterface>('v2_Release', ReleaseSchema)

export default ReleaseModel
