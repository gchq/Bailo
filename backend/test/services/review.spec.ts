import { describe, expect, test, vi } from 'vitest'

import AccessRequest from '../../src/models/AccessRequest.js'
import Model from '../../src/models/Model.js'
import Release from '../../src/models/Release.js'
import {
  createAccessRequestReviews,
  createReleaseReviews,
  findReviews,
  findReviewsForAccessRequests,
  removeAccessRequestReviews,
} from '../../src/services/review.js'

vi.mock('../../src/connectors/authorisation/index.js')
vi.mock('../../src/connectors/authentication/index.js', async () => ({
  default: { getEntities: vi.fn(() => ['user:test']) },
}))

const reviewModelMock = vi.hoisted(() => {
  const obj: any = { kind: 'access', accessRequestId: 'Hello', role: 'msro' }

  obj.aggregate = vi.fn(() => obj)
  obj.match = vi.fn(() => obj)
  obj.sort = vi.fn(() => obj)
  obj.lookup = vi.fn(() => obj)
  obj.append = vi.fn(() => obj)
  obj.find = vi.fn(() => [obj])
  obj.findOne = vi.fn(() => obj)
  obj.findOneAndUpdate = vi.fn(() => obj)
  obj.findByIdAndUpdate = vi.fn(() => obj)
  obj.updateOne = vi.fn(() => obj)
  obj.save = vi.fn(() => obj)
  obj.delete = vi.fn(() => obj)
  obj.limit = vi.fn(() => obj)
  obj.unwind = vi.fn(() => obj)
  obj.at = vi.fn(() => obj)
  obj.map = vi.fn(() => [])
  obj.filter = vi.fn(() => [])

  const model: any = vi.fn(() => obj)
  Object.assign(model, obj)

  return model
})

vi.mock('../../src/models/Review.js', async () => ({
  ...((await vi.importActual('../../src/models/Review.js')) as object),
  default: reviewModelMock,
}))

const smtpMock = vi.hoisted(() => ({
  notifyReviewResponseForAccess: vi.fn(() => Promise.resolve()),
  notifyReviewResponseForRelease: vi.fn(() => Promise.resolve()),
  requestReviewForRelease: vi.fn(() => Promise.resolve()),
  requestReviewForAccessRequest: vi.fn(() => Promise.resolve()),
}))
vi.mock('../../src/services/smtp/smtp.js', async () => smtpMock)

const modelMock = vi.hoisted(() => ({
  getModelById: vi.fn(),
}))
vi.mock('../../src/services/model.js', async () => modelMock)

const logMock = vi.hoisted(() => ({
  info: vi.fn(),
  warn: vi.fn(),
  error: vi.fn(),
}))
vi.mock('../../src/services/log.js', async () => ({
  default: logMock,
}))
const arrayUtilMock = vi.hoisted(() => ({
  asyncFilter: vi.fn(),
}))
vi.mock('../../src/utils/array.js', async () => arrayUtilMock)

describe('services > review', () => {
  const user: any = { dn: 'test' }

  test('findReviews > all reviews for user', async () => {
    await findReviews(user, true)

    expect(reviewModelMock.match.mock.calls.at(0)).toMatchSnapshot()
    expect(reviewModelMock.match.mock.calls.at(1)).toMatchSnapshot()
  })

  test('findReviews > active reviews for a specific model', async () => {
    await findReviews(user, false, 'modelId')

    expect(reviewModelMock.match.mock.calls.at(0)).toMatchSnapshot()
    expect(reviewModelMock.match.mock.calls.at(1)).toMatchSnapshot()
  })

  test('findReviewsForAccessRequests > success', async () => {
    await findReviewsForAccessRequests([reviewModelMock.accessRequestId])

    expect(reviewModelMock.find.mock.calls.at(0)).toMatchSnapshot()
  })

  test('createReleaseReviews > No entities found for required roles', async () => {
    const result: Promise<void> = createReleaseReviews(new Model(), new Release())

    expect(smtpMock.requestReviewForRelease).not.toBeCalled()
    expect(result).resolves.not.toThrowError()
    expect(reviewModelMock.save).toBeCalled()
  })

  test('createReleaseReviews > successful', async () => {
    await createReleaseReviews(
      new Model({ collaborators: [{ entity: 'user:user', roles: ['msro', 'mtr'] }] }),
      new Release(),
    )

    expect(reviewModelMock.save).toBeCalled()
    expect(smtpMock.requestReviewForRelease).toBeCalled()
  })

  test('createAccessRequestReviews > successful', async () => {
    await createAccessRequestReviews(
      new Model({ collaborators: [{ entity: 'user:user', roles: ['msro', 'mtr'] }] }),
      new AccessRequest(),
    )

    expect(reviewModelMock.save).toBeCalled()
    expect(smtpMock.requestReviewForAccessRequest).toBeCalled()
  })

  test('removeAccessRequestReviews > successful', async () => {
    await removeAccessRequestReviews(reviewModelMock.accessRequestId)

    expect(reviewModelMock.find.mock.calls.at(0)).toMatchSnapshot()
    expect(reviewModelMock.delete).toBeCalled()
  })

  test('removeAccessRequestReviews > could not delete failure', async () => {
    reviewModelMock.delete.mockImplementationOnce(() => {
      throw Error('Error deleting object')
    })

    expect(() => removeAccessRequestReviews('')).rejects.toThrowError(
      /^The requested access request review could not be deleted./,
    )
  })
})
