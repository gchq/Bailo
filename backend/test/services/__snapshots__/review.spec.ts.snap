// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`services > review > findReviews > active reviews for a specific model 1`] = `
[
  {
    "modelId": "modelId",
  },
]
`;

exports[`services > review > findReviews > active reviews for a specific model 2`] = `
[
  {},
]
`;

exports[`services > review > findReviews > all reviews for user 1`] = `
[
  {},
]
`;

exports[`services > review > findReviews > all reviews for user 2`] = `
[
  {
    "$expr": {
      "$gt": [
        {
          "$size": {
            "$filter": {
              "as": "item",
              "cond": {
                "$and": [
                  {
                    "$in": [
                      "$$item.entity",
                      [
                        "user:test",
                      ],
                    ],
                  },
                  {
                    "$in": [
                      "$role",
                      "$$item.roles",
                    ],
                  },
                ],
              },
              "input": "$model.collaborators",
            },
          },
        },
        0,
      ],
    },
  },
]
`;

exports[`services > review > findReviewsForAccessRequests > success 1`] = `
[
  {
    "accessRequestId": [
      "Hello",
    ],
  },
]
`;

exports[`services > review > removeAccessRequestReviews > successful 1`] = `
[
  {
    "accessRequestId": [
      "Hello",
    ],
  },
]
`;
